system/jhtree/jhtree.cpp

```diff
@@ -2334,134 +2357,106 @@ extern jhtree_decl void getNodeCacheInfo(ICacheInfoRecorder &cacheInfo)
+        (*addMetric[cacheType])++;
+        cache[cacheType].add(key, *LINK(node));
+        return node;
     }
-    CJHTreeNode *node;
-    {
-        CriticalUnblock block(lock);
-        node = keyIndex->loadNode(pos);  // NOTE - don't want cache locked while we load!
-    }
-    cacheAdds++;
-    switch (cacheType)
+    else
     {
-    case NodeBranch:
-        if (cacheNodes)
-        {
-            CJHTreeNode *cacheNode = nodeCache.query(key); // check if added to cache while we were reading
-            if (cacheNode)
-            {
-                ::Release(node);
-                cacheHits++;
-                if (ctx) ctx->noteStatistic(StNumNodeCacheHits, 1);
-                nodeCacheHits++;
-                return LINK(cacheNode);
-            }
-            if (ctx) ctx->noteStatistic(StNumNodeCacheAdds, 1);
-            nodeCacheAdds++;
-            nodeCache.add(key, *LINK(node));
-        }
-        break;
-    case NodeLeaf:
-        if (cacheLeaves)
+        CJHTreeNode * node;
         {
-            CJHTreeNode *cacheNode = leafCache.query(key); // check if added to cache while we were reading
-            if (cacheNode)
+            CriticalBlock block(lock);
+
+            node = cache[cacheType].query(key);
+            if (likely(node))
             {
-                ::Release(node);
                 cacheHits++;
-                if (ctx) ctx->noteStatistic(StNumLeafCacheHits, 1);
-                leafCacheHits++;
-                return LINK(cacheNode);
+                if (ctx) ctx->noteStatistic(hitStatId[cacheType], 1);
+                (*hitMetric[cacheType])++;
             }
-            if (ctx) ctx->noteStatistic(StNumLeafCacheAdds, 1);
-            leafCacheAdds++;
-            leafCache.add(key, *LINK(node));
-        }
-        break;
-    case NodeBlob:
-        if (cacheBlobs)
-        {
-            CJHTreeNode *cacheNode = blobCache.query(key); // check if added to cache while we were reading
-            if (cacheNode)
+            else
             {
-                ::Release(node);
-                cacheHits++;
-                if (ctx) ctx->noteStatistic(StNumBlobCacheHits, 1);
-                blobCacheHits++;
-                return LINK(cacheNode);
+                node = keyIndex->createNode(type);
+                cacheAdds++;
+                if (ctx) ctx->noteStatistic(addStatId[cacheType], 1);
+                (*addMetric[cacheType])++;
+                cache[cacheType].add(key, *node);
             }
-            if (ctx) ctx->noteStatistic(StNumBlobCacheAdds, 1);
-            blobCacheAdds++;
-            blobCache.add(key, *LINK(node));
+            node->Link();
         }
-        break;
+
+        //Actually load the information outside the critical section
+        if (likely(node->isReady()))
+            return node;
+
+        //Reather than having a critical section in each node, use an array indexed by a hash of the node address
```

Comment:
Typo Reather

Reply:
It's a shame to recalculate the hash that we presumably just calculated above to look up in the cache. Maybe not a big deal...

Reply:
It's a shame to recalculate the hash that we presumably just calculated above to look up in the cache. Maybe not a big deal...